aws/resource_aws_dms_endpoint.go

```diff
@@ -238,26 +284,50 @@ func resourceAwsDmsEndpointUpdate(d *schema.ResourceData, meta interface{}) erro
 		hasChanges = true
 	}
 
-	if d.HasChange("service_access_role") {
-		request.DynamoDbSettings = &dms.DynamoDbSettings{
-			ServiceAccessRoleArn: aws.String(d.Get("service_access_role").(string)),
-		}
-		hasChanges = true
-	}
-
-	if d.HasChange("endpoint_type") {
-		request.EndpointType = aws.String(d.Get("endpoint_type").(string))
-		hasChanges = true
-	}
-
 	if d.HasChange("engine_name") {
 		request.EngineName = aws.String(d.Get("engine_name").(string))
 		hasChanges = true
 	}
 
-	if d.HasChange("extra_connection_attributes") {
-		request.ExtraConnectionAttributes = aws.String(d.Get("extra_connection_attributes").(string))
-		hasChanges = true
+	switch d.Get("engine_name").(string) {
+	case "dynamodb":
+		if d.HasChange("service_access_role") {
+			request.DynamoDbSettings = &dms.DynamoDbSettings{
+				ServiceAccessRoleArn: aws.String(d.Get("service_access_role").(string)),
+			}
+			hasChanges = true
+		}
+	case "s3":
+		if d.HasChange("service_access_role") || d.HasChange("bucket_name") || d.HasChange("bucket_folder") || d.HasChange("extra_connection_attributes") {
+			request.S3Settings = &dms.S3Settings{
+				ServiceAccessRoleArn: aws.String(d.Get("service_access_role").(string)),
+				BucketFolder:         aws.String(d.Get("bucket_folder").(string)),
+				BucketName:           aws.String(d.Get("bucket_name").(string)),
+			}
+
+			elems := strings.Split(d.Get("extra_connection_attributes").(string), ";")
+			if len(elems) > 0 {
+				for _, elem := range elems {
+					vals := strings.Split(elem, "=")
+					if strings.Contains(strings.ToLower(vals[0]), "compressiontype") {
+						request.S3Settings.CompressionType = aws.String(vals[1])
+					} else if strings.Contains(strings.ToLower(vals[0]), "csvdelimiter") {
+						request.S3Settings.CsvDelimiter = aws.String(vals[1])
+					} else if strings.Contains(strings.ToLower(vals[0]), "csvrowdelimiter") {
+						request.S3Settings.CsvRowDelimiter = aws.String(vals[1])
+					}
+				}
+			}
+
+			hasChanges = true
+
+			goto DONE
```

Comment:
Can we please implement this without `goto`? Even if that means duplicated code.

Reply:
Thanks for the feedback @bflad 
Ok I will implement without goto. 
Usually I also don't like goto approach, but usually as I thought it is normal in GO 👍
Will fix it and other feedback right away